src/spi.rs

```diff
@@ -414,14 +441,62 @@ impl<'d> SpiMasterDriver<'d> {
             quadwp_io_num: -1,
             quadhd_io_num: -1,
 
-            max_transfer_sz: config.dma.max_transfer_size() as i32,
+            max_transfer_sz: max_transfer_sz as i32,
             ..Default::default()
         };
 
-        esp!(unsafe { spi_bus_initialize(SPI::device(), &bus_config, config.dma.into()) })?;
+        esp!(unsafe { spi_bus_initialize(SPI::device(), &bus_config, dma_chan) })?;
+        Ok(max_transfer_sz)
+    }
+}
 
-        let device_config = spi_device_interface_config_t {
-            spics_io_num: cs.as_ref().map_or(-1, |p| p.pin()),
+impl<'d> Drop for SpiMasterDriver<'d> {
+    fn drop(&mut self) {
+        esp!(unsafe { spi_bus_free(self.host()) }).unwrap();
+    }
+}
+
+pub struct EspSpiDevice<'d, T> {
```

Comment:
i see that now. but if we compare it with the others devices. than maybe the SpiMasterDriver should be renamed to something without the *Driver. Technically you can not drive anything now with the SpiMasterDriver, only the Devices do that. Maybe something like SpiMasterInterface. Then we can Call the EspSpiDevice something like an SpiDeviceDriver. (like the driver for every gpio )Would this be more in line with the rest? + for shorter names :smile:  Sane would than go for the internal SpiBusMasterInterface maybe.

Reply:
`SpiMasterDriver` is the thing that initializes the SPI bus - in **master** mode - for the particular SPI peripheral. And keeps it initialized until drop. So in a way, it **is** a driver, in that to know that you have the bus initialized, you need to keep a reference to this "driver" around. So:
- We need the `Master` word as in future we'll have the same notion (bus initialized) but for the slave drivers
- We need `Spi` (obviously)
- Do we need `Driver`? If not, then we are back into the current release havoc, where it is not easy to tell if something is a driver, or a peripheral, or a supporting structure

Besides `SpiMasterDriver`, sone additional suggestions:
- `SpiBusMasterDriver` - name collides with the thing you get in the callback provided in `transaction`
- `SpiHostMasterDriver ` - introduces the "host" word though

I think the current name is short enough, honestly...

> Then we can Call the EspSpiDevice something like an SpiDeviceDriver

But that's also what *I* suggest, except that you *have to* have the `Master` word there as well, because someday we'll also expose the ESP IDF slave SPI API. Look at how the i2c driver uses the master/slave dichotomy for more details.